extensions/flight-sql/src/main/java/io/deephaven/server/flightsql/FlightSqlActionHelper.java

```diff
@@ -0,0 +1,196 @@
+
+        T visit(ActionEndTransactionRequest action);
+
+        T visit(@SuppressWarnings("deprecation") ActionCancelQueryRequest action);
+
+        T visit(ActionCreatePreparedSubstraitPlanRequest action);
+    }
+
+    public static boolean handlesAction(String type) {
+        // There is no prefix for Flight SQL action types, so the best we can do is a set-based lookup. This also means
+        // that this resolver will not be able to respond with an appropriately scoped error message for new Flight SQL
+        // action types (io.deephaven.server.flightsql.FlightSqlResolver.UnsupportedAction).
+        return FLIGHT_SQL_ACTION_TYPES.contains(type);
+    }
+
+    public static <T> T visit(Action action, ActionVisitor<T> visitor) {
+        final String type = action.getType();
+        switch (type) {
+            case CREATE_PREPARED_STATEMENT_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionCreatePreparedStatementRequest.class));
+            case CLOSE_PREPARED_STATEMENT_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionClosePreparedStatementRequest.class));
+            case BEGIN_SAVEPOINT_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionBeginSavepointRequest.class));
+            case END_SAVEPOINT_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionEndSavepointRequest.class));
+            case BEGIN_TRANSACTION_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionBeginTransactionRequest.class));
+            case END_TRANSACTION_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionEndTransactionRequest.class));
+            case CANCEL_QUERY_ACTION_TYPE:
+                // noinspection deprecation
+                return visitor.visit(unpack(action.getBody(), ActionCancelQueryRequest.class));
+            case CREATE_PREPARED_SUBSTRAIT_PLAN_ACTION_TYPE:
+                return visitor.visit(unpack(action.getBody(), ActionCreatePreparedSubstraitPlanRequest.class));
+        }
+        // Should not get here unless handlesAction is implemented incorrectly.
+        throw new IllegalStateException(String.format("Unexpected Flight SQL Action type '%s'", type));
+    }
+
+    private static <T extends com.google.protobuf.Message> T unpack(byte[] body, Class<T> clazz) {
+        final Any any = parseActionOrThrow(body);
+        return unpackActionOrThrow(any, clazz);
+    }
+
+    private static Optional<Any> parse(byte[] data) {
+        try {
+            return Optional.of(Any.parseFrom(data));
+        } catch (final InvalidProtocolBufferException e) {
+            return Optional.empty();
+        }
+    }
+
+    private static Any parseActionOrThrow(byte[] data) {
+        return parse(data)
+                .orElseThrow(() -> FlightSqlErrorHelper.error(Status.Code.INVALID_ARGUMENT, "Invalid action"));
+    }
```

Comment:
This seems silly - it is private, it is only called in one place, and that caller just re-unwraps the optional or throws. 

```suggestion
    private static Any parseActionOrThrow(byte[] data) {
        try {
            return Any.parseFrom(data);
        } catch (final InvalidProtocolBufferException e) {
            throw FlightSqlErrorHelper.error(Status.Code.INVALID_ARGUMENT, "Invalid action");
        }
    }
```